README.md
[![Build Status](https://travis-ci.org/mcmillhj/Net-Easypost.svg?branch=master)](https://travis-ci.org/mcmillhj/Net-Easypost)
[![Coverage Status](https://coveralls.io/repos/mcmillhj/Net-Easypost/badge.svg?branch=master)](https://coveralls.io/r/mcmillhj/Net-Easypost?branch=master)
[![Cpan license](https://img.shields.io/cpan/l/Net-Easypost.svg)](https://metacpan.org/release/Net-Easypost)
[![Cpan version](https://img.shields.io/cpan/v/Net-Easypost.svg)](https://metacpan.org/release/Net-Easypost)

# NAME

Net::Easypost - Perl client for the Easypost web service

# VERSION

version 0.21

# SYNOPSIS

    use Net::Easypost;

    my $to = Net::Easypost::Address->new(
       name    => 'Johnathan Smith',
       street1 => '710 East Water Street',
       city    => 'Charlottesville',
       state   => 'VA',
       zip     => '22902',
       phone   => '(434)555-5555',
       country => 'US',
    );

    my $from = Net::Easypost::Address->new(
       name    => 'Jarrett Streebin',
       phone   => '3237078576',
       city    => 'Half Moon Bay',
       street1 => '310 Granelli Ave',
       state   => 'CA',
       zip     => '94019',
       country => 'US',
    );

    my $parcel = Net::Easypost::Parcel->new(
       length => 10.0,
       width  => 5.0,
       height => 8.0,
       weight => 10.0,
    );

    my $customs_info = Net::Easypost::CustomsInfo->new(
        customs_signer   => 'Steve Brule',
        contents_type    => 'merchandise',
        restriction_type => 'none',
        customs_certify  => 1,
        eel_ppc          => 'NOEEI 30.37(a)',
        customs_items => [
            Net::Easypost::CustomsItem->new(
                code             => '111111',
                description      => 'T-Shirt',
                quantity         => 1,
                weight           => 5,
                value            => 10,
                hs_tariff_number => '123456',
                origin_country   => 'US',
            )
        ]
    );

    my $shipment = Net::Easypost::Shipment->new(
        to_address   => $to,
        from_address => $from,
        parcel       => $parcel,
        customs_info => $customs_info,
    );

    my $ezpost = Net::Easypost->new;
    my $label = $ezpost->buy_label($shipment, ('rate' => 'lowest'));

    printf("You paid \$%0.2f for your label to %s\n", $label->rate->rate, $to);
    $label->save;
    printf "Your postage label has been saved to '" . $label->filename . "'\n";

# OVERVIEW

This is a Perl client for the postage API at [Easypost](https://www.easypost.com/docs/api). Consider this
API at beta quality mostly because some of these library calls have an inconsistent input
parameter interface which I'm not super happy about. Still, there's enough here to get
meaningful work done, and any future changes will be fairly cosmetic.

Please note! **All API errors are fatal via croak**. If you need to catch errors more gracefully, I
recommend using [Try::Tiny](https://metacpan.org/pod/Try::Tiny) in your implementation.

API key:

You must have your API key stored in an environment variable named
EASYPOST\_API\_KEY

# ATTRIBUTES

## requester

HTTP client to POST and GET

# METHODS

## verify\_address

This method attempts to validate an address. This call expects to take the same parameters
(in a hashref) or an instance of [Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address), namely:

- street1
- street2
- city
- state
- zip
- country

You may omit some of these attributes like city, state if you supply a zip, or
zip if you supply a city, state.

This call returns a new [Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address) object.

Along with the validated address, the `phone` and `name` fields will be
copied from the input parameters, if they're set.

The verification works only for addresses in US. If you pass a country
other than US (the default), a warning will be issued, but the
[Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address) object will be returned.

## get\_rates

This method will get postage rates between two zip codes. It takes the following input parameters:

- to => an instance of [Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address)
- from => an instance of [Net::Easypost::Address](https://metacpan.org/pod/Net::Easypost::Address)
- parcel => an instance of [Net::Easypost::Parcel](https://metacpan.org/pod/Net::Easypost::Parcel)

This call returns an array of [Net::Easypost::Rate](https://metacpan.org/pod/Net::Easypost::Rate) objects in an arbitrary order.

## buy\_label

This method will attempt to purchase postage and generate a shipping label.

It takes as input:

- A [Net::Easypost::Shipment](https://metacpan.org/pod/Net::Easypost::Shipment) object
- A [Net::Easypost::Rate](https://metacpan.org/pod/Net::Easypost::Rate) object

It returns a [Net::Easypost::Label](https://metacpan.org/pod/Net::Easypost::Label) object.

## get\_label

This method retrieves a label from a past purchase. It takes the label filename as its
only input parameter. It returns a [Net::Easypost::Label](https://metacpan.org/pod/Net::Easypost::Label) object.

## list\_labels

This method returns an arrayref with all past purchased label filenames. It takes no
input parameters.

# SUPPORT

Please report any bugs or feature requests to [https://github.com/mcmillhj/Net-Easypost/issues](https://github.com/mcmillhj/Net-Easypost/issues)

# SEE ALSO

- [Easypost API docs](https://www.easypost.com/docs/api)

# AUTHOR

Mark Allen <mrallen1@yahoo.com>, Hunter McMillen <mcmillhj@gmail.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Change history for Net-Easypost

{{$NEXT}}

0.21        2017-09-06
            - Contribution from @fendent that fixes an issue when creating CustomsInfo objects

0.20        2017-05-15
            - Add test dependency for Test::Exception
            - Add JSON::PP::Boolean type to all boolean Moo attributes to avoid cpantesters failure

0.19        2017-05-13
            - Add support for CustomsInfo & CustomsItem object to support international shipping
            - Remove default Country of US
            - Updated Net::Easypost::Address to support all fields returned in the API
            - Address verification is currently broken; apparently it hasn't worked since v2 of the Easypost API was released ... oops.

0.18        2016-01-05
            - Add JSON as dependency in dist.ini
0.17        2016-01-05
            - Remove Mojo::UserAgent in favor of HTTP::Tiny
            
0.16        2015-12-29
            - Minor changes, mostly formatting and build configuration
            - Now have an explicit cpanfile and Makefile.PL in the release in case some don't want to use Dist::Zilla
0.15        2015-03-18

0.14        2015-03-04

0.13        2014-05-27

0.12        2014-05-22

0.11        2014-05-21

0.09        2013-07-31
            Don't be so clever when passed a hashref in get_rates
            (Fixes RT#87493)
            Fix failing tests in t/01-basic.t
            Add test for RT#87493
            Add CPANfile to dist.ini

0.08        2013-05-19
            Seriously. Fix rate number test.

0.07        2013-05-16
            Check carriers against a regex
            For reals, make IO::Socket::SSL 
                a dependency

0.06        2013-05-15
            Make IO::Socket::SSL a prereq
            Check rate against regex

0.05        2013-05-14
            Fix post request method due to 
                Mojo deprecation warnings

0.04        2013-05-13
            Changed abstract
            Fixed dependency (Thanks Graham Knop)
            Fixed t/01-basic.t

0.03        2013-01-19
            Update URLs in POD
            Removed Data::Printer
            Return actual endpoint error codes 
            instead of inspecting JSON.

0.02        2013-01-13
            New endpoint
            Updated test suite

0.01        2012-12-04
            Initial release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.010.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl client for the Easypost web service",
  "AUTHOR" => "Mark Allen <mrallen1\@yahoo.com>, Hunter McMillen <mcmillhj\@gmail.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Net-Easypost",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Net::Easypost",
  "PREREQ_PM" => {
    "HTTP::Tiny" => "0.054",
    "IO::Socket::SSL" => "1.962",
    "JSON::MaybeXS" => "1.003009",
    "Mojolicious" => "4.66",
    "Moo" => "1.004002",
    "Types::Standard" => "0.040",
    "namespace::autoclean" => "0.28"
  },
  "TEST_REQUIRES" => {
    "Test::Exception" => 0
  },
  "VERSION" => "0.20",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "HTTP::Tiny" => "0.054",
  "IO::Socket::SSL" => "1.962",
  "JSON::MaybeXS" => "1.003009",
  "Mojolicious" => "4.66",
  "Moo" => "1.004002",
  "Test::Exception" => 0,
  "Types::Standard" => "0.040",
  "namespace::autoclean" => "0.28"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Net/Easypost/Address.pm
package Net::Easypost::Address;

use Carp qw/croak/;
use JSON::MaybeXS;
use Scalar::Util;
use overload
    '""'     => sub { $_[0]->as_string },
    '0+'     => sub { Scalar::Util::refaddr($_[0]) },
    fallback => 1;
use Types::Standard qw(Bool Enum HashRef InstanceOf Str Undef);


use Moo;
with qw/Net::Easypost::PostOnBuild/;
with qw/Net::Easypost::Resource/;
use namespace::autoclean;

has [qw/street1
     street2
     city
     state
     zip
     country
     carrier_facility
     name
     company
     phone
     email
     federal_tax_id
     state_tax_id/
] => (
    is  => 'rw',
    isa => Str|Undef
);

has 'residential' => (
    is     => 'rw',
    isa    => Bool|InstanceOf['JSON::PP::Boolean'],
    coerce => sub { $_[0] ? JSON->true : JSON->false }
);

has 'verifications' => (
    is  => 'rw',
    isa => HashRef
);

has 'verify_adress' => (
    is  => 'rw',
    isa => Enum[qw/zip4 delivery/]
);

sub _build_fieldnames {
    return [
	qw/
	street1
	street2
	city
	state
	zip
	country
	residential
	carrier_facility
	name
	company
	phone
	email
	federal_tax_id
	state_tax_id
	/
    ];
}

sub _build_role { 'address' }
sub _build_operation { '/addresses' }

sub clone {
    my ($self) = @_;

    return Net::Easypost::Address->new(
       map  { $_ => $self->$_ }
       grep { defined $self->$_ } @{ $self->fieldnames }
    );
}

sub serialize {
   my ($self) = @_;

   # want a hashref of e.g., address[name] => foo from all defined attributes
   # verify[]=zip4|delivery - verify an address
   return {
       # (defined $self->verify_address ? "verify[]" => $self->verify : ()),
       (map  { $self->role . "[$_]" => $self->$_ }
       grep { defined $self->$_ } @{ $self->fieldnames })
   };
}

sub as_string {
    my ($self) = @_;

    return join "\n",
        (map  { $self->$_ }
            grep { defined $self->$_ } qw(name phone street1 street2)),
        join " ",
            (map  { $self->$_ }
                grep { defined $self->$_ } qw(city state zip));
}

sub merge {
    my ($self, $old, $fields) = @_;

    map { $self->$_($old->$_) }
        grep { defined $old->$_ }
            @$fields;

    return $self;
}

sub verify {
    my ($self) = @_;

    my $verify_response =
       $self->requester->get(
          join '/', $self->operation, $self->id, 'verify'
       );

    croak 'Unable to verify address, failed with message: '
        . $verify_response->{error}
    if $verify_response->{error};

    my $new_address = Net::Easypost::Address->new(
        $verify_response->{address}
    );

    return $new_address->merge($self, [qw(phone name)]);
}

1;

__END__

=pod

=head1 NAME

 Net::Easypost::Address

=head1 SYNOPSIS

 Net::Easypost::Address->new

=head1 ATTRIBUTES

=over 4

=item mode

 string: Set based on which api-key you used, either "test" or "production"

=item street1

 string: First line of the address

=item street2

 string: Second line of the address

=item city

 string: City the address is located in

=item state

 string: State or province the address is located in

=item zip

 string: ZIP or postal code the address is located in

=item country

 string: ISO 3166 country code for the country the address is located in

=item residential

 boolean: Whether or not this address would be considered residential

=item carrier_facility

 string: The specific designation for the address (only relevant if the address is a carrier facility)

=item name

 string: Name of the person. Both name and company can be included

=item company

 string: Name of the organization. Both name and company can be included

=item phone

 string: Phone number to reach the person or organization

=item email

 string: Email to reach the person or organization

=item federal_tax_id

 string: Federal tax identifier of the person or organization

=item state_tax_id

 string: State tax identifier of the person or organization

=item verifications

 *CURRENTLY NOT IMPLEMENTED* Verifications: The result of any verifications requested

=back

=head1 METHODS

=over 4

=item _build_fieldnames

Attributes that make up an Address, from L<Net::Easypost::Resource>

=item _build_role

Prefix to data when POSTing to the Easypost API about Address objects

=item _build_operation

Base API endpoint for operations on Address objects

=item clone

Make a new copy of this object and return it

=item as_string

Format this address as it might be seen on a mailing label. This class overloads
stringification using this method, so something like C<say $addr> should just work.

=item merge

This method takes a L<Net::Easypost::Address> object and an arrayref of fields to copy
into B<this> object. This method only merges fields that are defined on the other object.

=item verify

This method takes a L<Net::Easypost::Address> object and verifies its underlying
address.

If a non-US address is asked for verification, a warning will be
emitted and the object itself will be returned.

=back

=cut


lib/Net/Easypost/Request.pm
package Net::Easypost::Request;

use Moo;

use Carp qw(croak);
use HTTP::Tiny;
use JSON::MaybeXS qw(decode_json);

has 'user_agent' => (
    is      => 'ro',
    default => sub { HTTP::Tiny->new( agent => 'Net-Easypost ' ) },
);

has 'endpoint' => (
    is      => 'ro',
    default => 'api.easypost.com/v2',
);

sub post {
    my ($self, $operation, $params) = @_;

    my $http_response = $self->user_agent->post_form(
       $self->_build_url($operation), $params,
    );

    unless ( $http_response->{success} ) {
        my ($err, $code) = map { $http_response->{$_} } qw(reason response);
        croak $code
           ? "FATAL: " . $self->endpoint . $operation . " returned $code: '$err'"
           : "FATAL: " . $self->endpoint . $operation . " returned '$err'";
    }

    return decode_json $http_response->{content};
}

sub get {
    my ($self, $endpoint) = @_;

    $endpoint = $self->_build_url($endpoint)
        unless $endpoint =~ m/https?/;

    my $http_response = $self->user_agent->get( $endpoint );
    return lc $http_response->{headers}->{'content-type'} =~ m|^\Qapplication/json\E|
       ? decode_json $http_response->{content}
       : $http_response->{content};
}

sub _build_url {
    my ($self, $operation) = @_;

    return 'https://' . $ENV{EASYPOST_API_KEY} . ':@' . $self->endpoint . $operation
        if exists $ENV{EASYPOST_API_KEY};

    croak 'Cannot find API key in access_code attribute of Net::Easypost'
        . ' or in an environment variable name EASYPOST_API_KEY';
}

1;

__END__

=pod

=head1 NAME

Net::Easypost::Request

=head1 SYNOPSIS

Net::Easypost::Request->new

=head1 ATTRIBUTES

=over 4

=item user_agent

A user agent attribute. Defaults to L<Mojo::UserAgent>.

=item endpoint

The Easypost service endpoint. Defaults to 'https://api.easypost.com/v2'

=back

=head1 METHODS

=over 4

=item _build_url

Given an operation, constructs a valid Easypost URL using the specified
EASYPOST_API_KEY

=item post

This method uses the C<user_agent> attribute to generate a form post request. It takes
an endpoint URI fragment and the parameters to be sent.  It returns JSON deserialized
into Perl structures.

=item get

This method uses the C<user_agent> attribute to generate a GET request to an endpoint. It
takes a complete endpoint URI as its input and returns a L<Mojo::Message::Response>
object.

=back

=cut


lib/Net/Easypost/Shipment.pm
package Net::Easypost::Shipment;

use Carp qw/croak/;
use Types::Standard qw/ArrayRef HashRef InstanceOf Str/;

use Moo;
with qw/Net::Easypost::PostOnBuild/;
with qw/Net::Easypost::Resource/;
use namespace::autoclean;

has [qw/from_address to_address/] => (
   is       => 'ro',
   isa      => InstanceOf['Net::Easypost::Address'],
   required => 1,
);

has 'parcel' => (
   is       => 'rw',
   isa      => InstanceOf['Net::Easypost::Parcel'],
   required => 1,
);

has 'customs_info' => (
   is  => 'rw',
   isa => InstanceOf['Net::Easypost::CustomsInfo'],
);

has 'scan_form' => (
   is  => 'rw',
   isa => InstanceOf['Net::Easypost::ScanForm'],
);

has 'rates' => (
   is  => 'rwp',
   isa => ArrayRef[InstanceOf['Net::Easypost::Rate']],
);

has 'options' => (
   is  => 'rw',
   isa => HashRef[Str],
);

sub _build_fieldnames {
    return [qw/to_address from_address parcel customs_info scan_form rates options/];
}

sub _build_role      { 'shipment'   }
sub _build_operation { '/shipments' }

sub BUILD {}
after BUILD => sub {
   my ($self) = @_;

   my $resp = $self->requester->post(
      $self->operation,
      $self->serialize,
   );
   $self->_set_id( $resp->{id} );
   $self->_set_rates(
      [  map {
            Net::Easypost::Rate->new(
               id          => $_->{id},
               carrier     => $_->{carrier},
               service     => $_->{service},
               rate        => $_->{rate},
               shipment_id => $self->id,
            )
         } @{ $resp->{rates} }
      ]
   );
};

sub serialize {
   my ($self) = @_;

   # want a hashref of e.g., shipment[to_address][id] => foo from all defined attributes
   return {
       map { $self->role . "[$_][id]" => $self->$_->id }
       grep { defined $self->$_ }
          qw(to_address from_address parcel customs_info)
   };
}

sub clone {
   my ($self) = @_;

   return Net::Easypost::Shipment->new(
       map  { $_ => $self->$_ }
       grep { defined $self->$_ }
          'id', @{ $self->fieldnames }
   );
}

sub buy {
   my ($self, %options) = @_;

   my $rate;
   if (exists $options{rate} && $options{rate} eq 'lowest') {
      ($rate) =
         sort { $a->{rate} <=> $b->{rate} } @{ $self->rates };
   }
   elsif (exists $options{service_type}) {
      ($rate) =
         grep { $options{service_type} eq $_->service } @{ $self->rates };
   }
   else {
      croak "Missing 'service' or 'rate' from options hash";
   }

   unless ($rate) {
      my $msg = "Allowed services and rates for this shipment are:\n";
      foreach my $rate (@{ $self->rates }) {
         $msg .= sprintf("\t%-15s: %4.2f\n", $rate->service, $rate->rate);
      }

      croak "Invalid service '$options{service_type}' selected for shipment " . $self->id . "\n$msg";
   }

   my $response = $self->requester->post(
      $self->operation . '/' . $self->id . '/buy',
      $rate->serialize
   );

   my $label = $response->{postage_label};
   return Net::Easypost::Label->new(
      id            => $label->{id},
      tracking_code => $response->{tracking_code},
      url           => $label->{label_url},
      filetype      => $label->{label_file_type},
      rate          => Net::Easypost::Rate->new($response->{selected_rate}),
      filename      => 'EASYPOST_LABEL_'
                        . $label->{id}
                        . '.'
                        . substr($label->{label_file_type}, index($label->{label_file_type}, '/') + 1),
   );
}

1;

__END__

=pod

=head1 NAME

Net::Easypost::Shipment

=head1 SYNOPSIS

Net::Easypost::Shipment->new

=head1 ATTRIBUTES

=over 4

=item to_address

Address this Shipment is going to

=item from_address

Address this Shipment is coming from

=item parcel

Parcel that is being shipped

=item customs_info

Customs information about this Shipment, optional (only for international shipments)

=item scan_form

USPS Tracking information, optional

=item rates

Array of Net::Easypost::Rate objects

=item options

Array of shipping options, may not be supported by all carriers

=back

=head1 METHODS

=over 4

=item _build_fieldnames

Attributes that make up an Address, from L<Net::Easypost::Resource>

=item _build_role

Prefix to data when POSTing to the Easypost API about Address objects

=item _build_operation

Base API endpoint for operations on Address objects

=item BUILD

Constructor for a Shipment object, overrides the base BUILD method in L<Net::Easypost::PostOnBuild>

=item clone

return a deep-copy of this object

=item serialize

Format this object into a form suitable for use with the Easypost service.
Overrides base implementation in L<Net::Easypost::PostOnBuild>

=item buy


=back

=cut


lib/Net/Easypost/CustomsInfo.pm
package Net::Easypost::CustomsInfo;

use JSON::MaybeXS;
use Types::Standard qw(ArrayRef Bool Enum InstanceOf Str);

use Moo;
with qw/Net::Easypost::Resource/;
with qw/Net::Easypost::PostOnBuild/;
use namespace::autoclean;

has 'eel_ppc' => (
    is  => 'rw',
);

has 'contents_type' => (
    is  => 'rw',
    isa => Enum[qw/documents gift merchandise returned_goods sample other/]
);

has 'customs_certify' => (
    is  => 'rw',
    isa => Bool|InstanceOf['JSON::PP::Boolean'],
    coerce => sub { $_[0] ? JSON->true : JSON->false }
);

has 'non_delivery_option' => (
    is      => 'rw',
    isa     => Enum[qw/abandon return/],
    default => 'return'
);

has [qw/contents_explanation customs_signer/] => (
    is  => 'rw',
    isa => Str
    );

has 'restriction_type' => (
    is  => 'rw',
    isa => Enum[qw/none quarantine sanitary_phytosanitary_inspection/]
);

has 'restriction_comments' => (
    is  => 'rw',
    isa => Str
);

has 'customs_items' => (
    is  => 'rw',
    isa => ArrayRef[InstanceOf['Net::Easypost::CustomsItem']]
);

sub _build_fieldnames {
    return [
	qw/
	contents_explanation
	contents_type
	customs_certify
	customs_signer
	eel_ppc
	non_delivery_option
	restriction_comments
	restriction_type
	/
    ];
}
sub _build_role { 'customs_info' }
sub _build_operation { '/customs_infos' }

sub serialize {
   my ($self) = @_;

   my $obj = {
       map  { $self->role . "[$_]" => $self->$_ }
       grep { defined $self->$_ } @{ $self->fieldnames }
   };

   # if customs_items exist, they were already created in Net::Easypost::CustomsItem
   # so we can simply pass the id
   if ($self->customs_items) {
       foreach my $i (0 .. $#{ $self->customs_items }) {
	   my $item = $self->customs_items->[$i];
	   $obj->{$self->role . "[customs_items][$i][id]"} = $item->id;
       }
   }

   return $obj;
}

sub clone {
    my ($self) = @_;

    return Net::Easypost::CustomsInfo->new(
        map  { $_ => $self->$_ }
	grep { defined $self->$_ } @{ $self->fieldnames },
	'customs_items' => [
	    map { $_->clone } $self->customs_items
	]
    );
}

1;

__END__

=pod

=head1 NAME

 Net::Easypost::CustomsInfo

=head1 SYNOPSIS

 Net::Easypost::CustomsInfo->new

=head1 ATTRIBUTES

=over 4

=item eel_pfc

 string: "EEL" or "PFC" value less than $2500: "NOEEI 30.37(a)"; value greater than $2500: see Customs Guide

=item contents_type

 string: "documents", "gift", "merchandise", "returned_goods", "sample", or "other"

=item contents_explanation

 string: Human readable description of content. Required for certain carriers and always required if contents_type is "other"

=item customs_certify

 boolean: Electronically certify the information provided

=item customs_signer

 string: Required if customs_certify is true

=item non_delivery_option

 string: "abandon" or "return", defaults to "return"

=item restriction_type

 string: "none", "other", "quarantine", or "sanitary_phytosanitary_inspection"

=item restriction_comments

 string: Required if restriction_type is not "none"

=item customs_items

 [L<Net::Easypost::CustomsItem>]: describes the products being shipped

=back

=cut


lib/Net/Easypost/CustomsItem.pm
package Net::Easypost::CustomsItem;

use Types::Standard qw/Num Str Undef/;

use Moo;
with qw/Net::Easypost::PostOnBuild/;
with qw/Net::Easypost::Resource/;
use namespace::autoclean;

has [qw/description hs_tariff_number origin_country code currency/] => (
    is  => 'rw',
    isa => Str|Undef
);

has [qw/quantity value weight/] => (
    is  => 'rw',
    isa => Num
);

sub _build_fieldnames {
    return [
	qw/
	code
	currency
	description
	hs_tariff_number
	origin_country
	quantity
	value
	weight
	/
    ];
}

sub _build_role { 'customs_item' }

sub _build_operation { '/customs_items' }

sub clone {
    my ($self) = @_;

    return Net::Easypost::CustomsItems->new(
       map  { $_ => $self->$_ }
       grep { defined $self->$_ } @{ $self->fieldnames }
    );
}

1;

__END__

=pod

=head1 NAME

 Net::Easypost::CustomsItem

=head1 SYNOPSIS

 Net::Easypost::CustomsItem->new

=head1 ATTRIBUTES

=over 4

=item description

 string: Required, description of item being shipped

=item quantity

 float: Required, greater than zero

=item value

 float (USD): Required, greater than zero, total value (unit value * quantity)

=item weight

 float (oz): Required, greater than zero, total weight (unit weight * quantity)

=item hs_tariff_number

 string: Harmonized Tariff Schedule, e.g. "6109.10.0012" for Men's T-shirts

=item code

 string: SKU/UPC or other product identifier

=item origin_country

 string: Required, 2 char country code

=item currency

 string: 3 char currency code, default USD

=back

=cut


lib/Net/Easypost/PostOnBuild.pm
package Net::Easypost::PostOnBuild;

use Moo::Role;

sub BUILD {}
after 'BUILD' => sub {
    my $self = shift;

   my $resp = $self->requester->post(
      $self->operation,
      $self->serialize,
   );
   $self->_set_id( $resp->{id} );
};

sub serialize {
   my ($self, $attrs) = @_;
   $attrs //= $self->fieldnames;

   # want a hashref of e.g., role[field1] => foo from all defined attributes
   return {
      map  { $self->role . "[$_]" => $self->$_ }
      grep { defined $self->$_ } @$attrs
   };
}

1;

__END__

=pod

=head1 NAME

=head1 SYNOPSIS

=head1 METHODS

=over 4

=item BUILD

After the Net::Easypost::Resource has been constructure, sends a POST to the Easypost
service for the type of Net::Easypost::Resource being constructed to get a valid ID

=item serialize

Format the defined attributes for a call to the Easypost service.
Takes an arrayref of attributes to serialize. Defaults to the C<fieldnames> attribute.

=back

=cut


lib/Net/Easypost.pm
package Net::Easypost;

use Carp qw(croak);

use Net::Easypost::Address;
use Net::Easypost::Label;
use Net::Easypost::Parcel;
use Net::Easypost::Rate;
use Net::Easypost::Request;
use Net::Easypost::Shipment;

use Moo;
use namespace::autoclean;

# ABSTRACT: Perl client for the Easypost web service

=head1 SYNOPSIS

   use Net::Easypost;

   my $to = Net::Easypost::Address->new(
      name    => 'Johnathan Smith',
      street1 => '710 East Water Street',
      city    => 'Charlottesville',
      state   => 'VA',
      zip     => '22902',
      phone   => '(434)555-5555',
      country => 'US',
   );

   my $from = Net::Easypost::Address->new(
      name    => 'Jarrett Streebin',
      phone   => '3237078576',
      city    => 'Half Moon Bay',
      street1 => '310 Granelli Ave',
      state   => 'CA',
      zip     => '94019',
      country => 'US',
   );

   my $parcel = Net::Easypost::Parcel->new(
      length => 10.0,
      width  => 5.0,
      height => 8.0,
      weight => 10.0,
   );

   my $customs_info = Net::Easypost::CustomsInfo->new(
       customs_signer   => 'Steve Brule',
       contents_type    => 'merchandise',
       restriction_type => 'none',
       customs_certify  => 1,
       eel_ppc          => 'NOEEI 30.37(a)',
       customs_items => [
	   Net::Easypost::CustomsItem->new(
	       code             => '111111',
	       description	=> 'T-Shirt',
	       quantity		=> 1,
	       weight		=> 5,
	       value		=> 10,
	       hs_tariff_number => '123456',
	       origin_country	=> 'US',
	   )
       ]
   );

   my $shipment = Net::Easypost::Shipment->new(
       to_address   => $to,
       from_address => $from,
       parcel       => $parcel,
       customs_info => $customs_info,
   );

   my $ezpost = Net::Easypost->new;
   my $label = $ezpost->buy_label($shipment, ('rate' => 'lowest'));

   printf("You paid \$%0.2f for your label to %s\n", $label->rate->rate, $to);
   $label->save;
   printf "Your postage label has been saved to '" . $label->filename . "'\n";

=head1 OVERVIEW

This is a Perl client for the postage API at L<Easypost|https://www.easypost.com/docs/api>. Consider this
API at beta quality mostly because some of these library calls have an inconsistent input
parameter interface which I'm not super happy about. Still, there's enough here to get
meaningful work done, and any future changes will be fairly cosmetic.

Please note! B<All API errors are fatal via croak>. If you need to catch errors more gracefully, I
recommend using L<Try::Tiny> in your implementation.

API key:

You must have your API key stored in an environment variable named
EASYPOST_API_KEY

=cut

=attr requester

HTTP client to POST and GET

=cut

has 'requester' => (
    is      => 'ro',
    lazy    => 1,
    default => sub { return Net::Easypost::Request->new }
);

=method verify_address

This method attempts to validate an address. This call expects to take the same parameters
(in a hashref) or an instance of L<Net::Easypost::Address>, namely:

=over

=item * street1

=item * street2

=item * city

=item * state

=item * zip

=item * country

=back

You may omit some of these attributes like city, state if you supply a zip, or
zip if you supply a city, state.

This call returns a new L<Net::Easypost::Address> object.

Along with the validated address, the C<phone> and C<name> fields will be
copied from the input parameters, if they're set.

The verification works only for addresses in US. If you pass a country
other than US (the default), a warning will be issued, but the
L<Net::Easypost::Address> object will be returned.


=cut

sub verify_address {
    my ($self, $params) = @_;

    if ( ref($params) eq 'HASH' ) {
        return Net::Easypost::Address->new( $params )->verify;
    }
    elsif ( ref($params) eq 'Net::Easypost::Address' ) {
        return $params->verify;
    }
    else {
        croak "verify_address expects either a hashref or an instance of Net::Easypost::Address\n";
    }
}

=method get_rates

This method will get postage rates between two zip codes. It takes the following input parameters:

=over

=item * to => an instance of L<Net::Easypost::Address>

=item * from => an instance of L<Net::Easypost::Address>

=item * parcel => an instance of L<Net::Easypost::Parcel>

=back

This call returns an array of L<Net::Easypost::Rate> objects in an arbitrary order.

=cut

sub get_rates {
    my $self = shift; # shift is important here

    my $params;
    if ( scalar @_ == 1 ) {
        if ( ref( $_[0] ) ne 'HASH' ) {
            croak 'get_rates expects a hashref not a '. ref($params) .'\n';
        }
        else {
            $params = shift;
        }
    }
    else {
        $params = { @_ };
    }

    return Net::Easypost::Shipment->new(
        to_address   => $params->{to},
        from_address => $params->{from},
        parcel       => $params->{parcel},
    )->rates;
}

=method buy_label

This method will attempt to purchase postage and generate a shipping label.

It takes as input:

=over

=item * A L<Net::Easypost::Shipment> object

=item * A L<Net::Easypost::Rate> object

=back

It returns a L<Net::Easypost::Label> object.

=cut

sub buy_label {
    my ($self, $shipment, %options) = @_;

    croak 'Buy label expects a parameter of type Net::Easypost::Shipment'
        unless $shipment || ref($shipment) ne 'Net::Easypost::Shipment';

    return $shipment->buy(%options);
}

=method get_label

This method retrieves a label from a past purchase. It takes the label filename as its
only input parameter. It returns a L<Net::Easypost::Label> object.

=cut

sub get_label {
    my ($self, $label_filename) = @_;

    my $resp = $self->requester->post(
        '/postage/get',
        { label_file_name => $label_filename }
    );

    return Net::Easypost::Label->new(
        rate          => Net::Easypost::Rate->new( $resp->{rate} ),
        tracking_code => $resp->{tracking_code},
        filename      => $resp->{label_file_name},
        filetype      => $resp->{label_file_type},
        url           => $resp->{label_url}
    );
}

=method list_labels

This method returns an arrayref with all past purchased label filenames. It takes no
input parameters.

=cut

sub list_labels {
    my ($self) = @_;

    my $resp = $self->requester->get(
        $self->requester->_build_url('/postage/list')
    );

    return $resp->{postages};
}

=head1 SUPPORT

Please report any bugs or feature requests to L<https://github.com/mcmillhj/Net-Easypost/issues>

=head1 SEE ALSO

=over

=item * L<Easypost API docs|https://www.easypost.com/docs/api>

=back

=cut

1;
